docs/architecture/adr-005-subscription.md

```diff
@@ -0,0 +1,225 @@
+
+`SubscriptionID => Subscription`
+
+```go
+// SubscriptionID is a unique identifier to a subscription struct
+// hash(Name+Address)
+type SubscriptionID []byte
+```
+
+```go
+// Subscription Metadata contains information necessary for processing subscriptions
+type Metadata struct {
+    Amount    sdk.Coins      // amount to be collected for each subscription period
+    Period    time.time      // duration of subscription period
+    Collector sdk.AccAddress // address that will collect subscription payments
+}
+```
+
+```go
+// Represents a user's subscription to a service
+type Subscription {
+    Name       string
+    Subscriber sdk.AccAddress
+    Limit      int64
+    LastPaid   time.Time
+    Active     bool
+}
+```
+
+Message Handling:
+
+Note much of this logic may be done inside a keeper. This is a outline of how msgs will be handled.
+
+```go
+func HandleCreateSubscriptionMsg(ctx sdk.Context, msg CreateSubscriptionMsg) {
+    if SubscriptionExists(msg.Name) {
+        return DuplicateSubscriptionErr
+    }
+    StoreMetadata(ctx, msg.Name, msg.Amount, msg.Period, msg.Collector) // Store metadata in store under key "Metadata:msg.Name"
+    InitializeDueQueue(ctx, msg.Name) // Initialize empty queue and store under key "DueQueue:Name"
+}
+```
+
+```go
+func HandleSubscribeMsg(ctx sdk.Context, msg SubscribeMsg) {
+    if !SubscriptionExists(msg.Name) {
+        return NoSuchSubscriptionErr
+    }
+
+    subscriptionID := hash(msg.Name|msg.Subscriber)
+    metadata := GetMetaData(ctx, msg.Name)
+    // subscribeMsg will pay for the first period
+    err := bank.Send(msg.Subscriber, metadata.Collector, metadata.Amount)
+    if err != nil {
+        return err
+    }
+
+    // check if user already has subscribe to this service
+    // if so renew the subscription
+    if subscription := GetUserSubscriptions(msg.Subscriber, msg.Name); subscription != nil {
+        if msg.Limit == -1 {
+            subscription.Limit = -1
+        } else {
+            subscription.Limit += msg.Limit
+        }
+        // set active to true
+        subscription.Active = true
+        StoreSubscription(subscriptionID, subscription)
+        return nil
+    }   
+    subscription := Subscription{
+        Name:       metadata.Name,
+        Subscriber: msg.Subscriber,
+        Limit:      msg.Limit,
+        LastPaid:   ctx.BlockTime,
+        Active:     true
+    }
+    StoreSubscription(subscriptionID, subscription)
+    // adds subscription to list of user subscriptions
+    AppendToUserSubscriptions(subscriptionID)
+    // push to back of DueQueue for this service
+    PushToBackOfDueQueue(msg.Name, subscriptionID)
+}
+```
+
+```go
+func HandleUnsubscribeMsg(ctx sdk.Context, msg UnsubscribeMsg) {
+    subscription := GetUserSubscriptions(msg.Subscriber, msg.Name)
+    if subscription == nil {
+        return nil
+    }
+    subscription.Active = false
+    subscription.Limit = 0 // to ensure that renewal updates limit correctly
```

Comment:
Should I be resetting Limit so that any future Renewal will define the new limit in the message?